dask/delayed.py

```diff
@@ -367,52 +391,63 @@ def _get_binary_operator(cls, op, inv=False):
-            except AttributeError:
-                name = '%s-%s' % (type(obj).__name__, tokenize(obj, pure=pure))
-        object.__setattr__(self, '_dasks', [{name: obj}])
-        object.__setattr__(self, 'pure', pure)
+def call_function(func, args, kwargs, pure=False, nout=None):
+    dask_key_name = kwargs.pop('dask_key_name', None)
+    pure = kwargs.pop('pure', pure)
 
-    def __setstate__(self, state):
-        self.__init__(*state)
-        return self
+    if dask_key_name is None:
+        name = '%s-%s' % (funcname(func), tokenize(func, *args,
+                                                   pure=pure, **kwargs))
+    else:
+        name = dask_key_name
 
-    def __getstate__(self):
-        return (self._data, self._key, self.pure)
+    args, dasks = unzip(map(to_task_dasks, args), 2)
+    if kwargs:
+        dask_kwargs, dasks2 = to_task_dasks(kwargs)
+        dasks = dasks + (dasks2,)
+        task = (apply, func, list(args), dask_kwargs)
+    else:
+        task = (func,) + args
+
+    dasks = flat_unique(dasks)
+    dasks.append({name: task})
+    return Delayed(name, dasks, n=nout)
+
+
+class DelayedLeaf(Delayed):
+    __slots__ = ('_obj', '_key', '_pure', '_nout')
+    def __init__(self, obj, key, pure=False, nout=None):
+        self._obj = obj
+        self._key = key
+        self._pure = pure
+        self._nout = nout
 
     @property
-    def _key(self):
-        return first(self._dasks[0])
+    def dask(self):
+        return {self._key: self._obj}
 
     @property
-    def _data(self):
-        return first(self._dasks[0].values())
+    def _dasks(self):
+        return [self.dask]
 
     def __call__(self, *args, **kwargs):
-        dask_key_name = kwargs.pop('dask_key_name', None)
-        pure = kwargs.pop('pure', self.pure)
+        return call_function(self._obj, args, kwargs,
+                             pure=self._pure, nout=self._nout)
 
-        if dask_key_name is None:
-            name = (funcname(self._data) + '-' +
-                    tokenize(self._key, *args, pure=pure, **kwargs))
-        else:
-            name = dask_key_name
 
-        args, dasks = unzip(map(to_task_dasks, args), 2)
-        if kwargs:
-            dask_kwargs, dasks2 = to_task_dasks(kwargs)
-            dasks = dasks + (dasks2,)
-            task = (apply, self._data, list(args), dask_kwargs)
-        else:
-            task = (self._data,) + args
+class DelayedAttr(Delayed):
+    __slots__ = ('_obj', '_attr', '_key')
+    def __init__(self, obj, attr, key):
+        self._obj = obj
+        self._attr = attr
+        self._key = key
 
-        dasks = flat_unique(dasks)
-        dasks.append({name: task})
-        return Delayed(name, dasks)
+    @property
+    def _dasks(self):
+        return [{self._key: (getattr, self._obj._key, self._attr)}] + self._obj._dasks
+
+    def __call__(self, *args, **kwargs):
+        return call_function(methodcaller(self._attr), (self._obj,) + args, kwargs)
```

Comment:
Makes the graphs resulting from method and attribute access cleaner. Also removes the need for the inlining optimization, as the graphs are created properly the first time. I think the class is simple enough to merit the addition.